File: core/src/ru/myitschool/dcrawler/ai/pathfinding/graph/CellNode.java
package ru.myitschool.dcrawler.ai.pathfinding.graph;

import ru.myitschool.dcrawler.dungeon.DungeonCell;

public class CellNode extends Node {

    private DungeonCell cell;

    public CellNode(int x, int y, DungeonCell cell) {
        super(x, y);
        this.cell = cell;
    }

    public DungeonCell getCell() {
        return cell;
    }

    @Override
    public boolean canConnect(Node node) {
        if (!(node instanceof CellNode)) {
            return false;
        }
        int distance = Math.abs(node.getX() - getX()) + Math.abs(node.getY() - getY());
        // this and other one should be ground tiles and they should be neighbours
        return getCell().getTile().isGroundTile() && ((CellNode) node).getCell().getTile().isGroundTile() && distance == 1;
    }

    @Override
    public int connectionCost(Node to) {
        return ((CellNode) to).getCell().getTile().getHardness();
    }

    @Override
    public boolean isReachable() {
        return cell.isOccupied();
    }
}

File: core/src/com/wolg_vlad/dcrawler/entities/ai/pathfinding/graph/CellNode.kt
package com.wolg_vlad.dcrawler.entities.ai.pathfinding.graph

import com.wolg_vlad.dcrawler.dungeon.DungeonCell

class CellNode(x: Int, y: Int, val cell: DungeonCell) : Node(x, y) {

    override fun canConnect(node: Node): Boolean {
        if (node !is CellNode) {
            return false
        }
        val distance = Math.abs(node.x - x) + Math.abs(node.y - y)
        // this and other one should be ground tiles and they should be neighbours
        return cell.tile.groundTile && node.cell.tile.groundTile && distance == 1
    }

    override fun connectionCost(to: Node?): Int {
        return (to as CellNode?)!!.cell.tile.hardness
    }

    override val isReachable: Boolean
        get() = !cell.isOccupied

}